File: 2022/d03.zig
const std = @import("std");

// does this work without ptr?
fn mark_items(items: *std.bit_set.StaticBitSet(53), str: []const u8) void {
    for (str) |c| {
        const idx = if (c <= @as(i32, 'Z'))
            c - @as(i32, 'A') + 26
        else
            c - @as(i32, 'a');
        items.set(@intCast(usize, idx));
    }
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const allocator = gpa.allocator();

    const file = try std.fs.cwd().openFile("d03.in", .{ .mode = .read_only });
    defer file.close();

    const contents = try file.reader().readAllAlloc(allocator, 10 * 1024 * 1024);
    defer allocator.free(contents);

    var split_iter = std.mem.split(u8, contents, "\n");
    var point_sum: u32 = 0;
    while (split_iter.next()) |line| {
        if (line.len == 0) continue;

        const half = line.len / 2;
        const compartment1 = line[0..half];
        const compartment2 = line[half..line.len];
        var items1 = std.bit_set.StaticBitSet(53).initEmpty();
        mark_items(&items1, compartment1);
        var items2 = std.bit_set.StaticBitSet(53).initEmpty();
        mark_items(&items2, compartment2);
        // result will be saved in first set
        items1.setIntersection(items2);
        // assumes there's only one DUPLICATE
        point_sum += 1 + @intCast(u32, (items1.findFirstSet() orelse @panic("No active bit")));
    }

    split_iter = std.mem.split(u8, contents, "\n");
    var common_point_sum: u32 = 0;
    while (true) {
        var items1 = std.bit_set.StaticBitSet(53).initEmpty();
        mark_items(&items1, split_iter.next() orelse break);
        var items2 = std.bit_set.StaticBitSet(53).initEmpty();
        mark_items(&items2, split_iter.next() orelse break);
        var items3 = std.bit_set.StaticBitSet(53).initEmpty();
        mark_items(&items3, split_iter.next() orelse break);

        // result will be saved in first set
        items1.setIntersection(items2);
        items1.setIntersection(items3);
        // assumes there's only one DUPLICATE
        common_point_sum += 1 + @intCast(
            u32,
            (items1.findFirstSet() orelse @panic("No active bit")),
        );
    }

    std.debug.print("Part1: Duplicate item prios are {}\n", .{point_sum});
    std.debug.print("Part2: Common items prios are {}\n", .{common_point_sum});
}

File: 2022/rust/src/bin/d03.rs
use std::fs;
use std::io;

fn mark_items(items: &mut [bool], s: &str) {
    for c in s.as_bytes() {
        let idx = if *c <= b'Z' { c - b'A' + 26 } else { c - b'a' };
        items[idx as usize] = true;
    }
}

pub fn main() -> Result<(), io::Error> {
    let contents = fs::read_to_string("../d03.in")?;

    let mut point_sum = 0;
    for line in contents.split('\n') {
        if line == "" { continue };
        // println!("{}", line);

        let half = line.len() / 2;
        let (compartment1, compartment2) = line.split_at(half);
        // no std bitset :/
        let mut items1 = [false; 53];
        mark_items(&mut items1[..], compartment1);
        let mut items2 = [false; 53];
        mark_items(&mut items2[..], compartment2);
        for i in 0..items1.len() {
            let in_both = items1[i] & items2[i];
            if in_both {
                point_sum += i + 1;
            }
        }
    }

    println!("Part1: Duplicate items prios are {}", point_sum);

    let mut common_point_sum = 0;
    let mut rucksacks = [
        [false; 53],
        [false; 53],
        [false; 53],
    ];
    // so we don't sum up instantly -> start at 1
    let mut idx = 1;
    for line in contents.split('\n') {
        if line == "" { continue };
        // println!("{}", line);

        let items = &mut rucksacks[idx % 3];
        mark_items(&mut items[..], line);

        if idx % 3 == 0 {
            for i in 0..rucksacks[0].len() {
                let in_all = rucksacks[0][i] & rucksacks[1][i] & rucksacks[2][i];
                if in_all {
                    common_point_sum += i + 1;
                }
            }
            // reset
            rucksacks = [
                [false; 53],
                [false; 53],
                [false; 53],
            ];
        }

        idx += 1;
    }

    println!("Part2: Common item prios are {}", common_point_sum);

    Ok(())
}
